lenovo/legion/16achg6/hybrid/default.nix

```diff
@@ -10,12 +10,14 @@
     ../../../../common/pc/laptop/ssd
   ];
 
-  services.xserver.videoDrivers = [ "nvidia" "amdgpu" ];
+  services.xserver.videoDrivers = [ "nvidia" ];
+  boot.initrd.kernelModules = [ "amdgpu" ];
```

Comment:
Apparently this works better when defined this way.
Someone recommended this approach to me on [nixos discourse](https://discourse.nixos.org/t/what-else-can-i-try-to-make-my-optimus-nvidia-card-work-well-with-x11/36958/7) and it did the trick.